> Other option let's make API auth failures not stop the program.

It doesn't. The error and fix suggestions are displayed; then, everything else is rendered modulo the cost estimates.

>  we might be able to prevent the new application-default login by having the program look for credentials in the environment first

I think we'll need to refactor roachprod permissions in a separate PR. Ideally, it should use a service account. For now, if `application-default` is not found, it errors out and suggests a fix,

```
04:19:47 gcloud.go:1041: Error during cost estimation (will continue without): google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
04:19:47 gcloud.go:1043: To fix this, run `gcloud auth application-default login`
```

--- a/go.mod
+++ b/go.mod
@@ -134,6 +134,7 @@ require (
 	github.com/elastic/gosigar v0.14.1
 	github.com/emicklei/dot v0.15.0
 	github.com/facebookgo/clock v0.0.0-20150410010913-600d898af40a
+	github.com/fatih/color v1.9.0
 	github.com/fraugster/parquet-go v0.10.0
 	github.com/fsnotify/fsnotify v1.5.1
 	github.com/getsentry/sentry-go v0.12.0
@@ -330,6 +331,7 @@ require (
 	github.com/lufia/plan9stats v0.0.0-20211012122336-39d0f177ccd0 // indirect
 	github.com/magiconair/properties v1.8.5 // indirect
 	github.com/mailru/easyjson v0.7.6 // indirect
+	github.com/mattn/go-colorable v0.1.12 // indirect
 	github.com/mattn/go-localereader v0.0.1 // indirect
 	github.com/mattn/go-runewidth v0.0.14 // indirect
 	github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369 // indirect

--- a/pkg/cmd/roachprod/BUILD.bazel
+++ b/pkg/cmd/roachprod/BUILD.bazel
@@ -22,9 +22,13 @@ go_library(
         "//pkg/roachprod/vm",
         "//pkg/roachprod/vm/gce",
         "//pkg/util/flagutil",
+        "//pkg/util/timeutil",
         "@com_github_cockroachdb_errors//:errors",
+        "@com_github_fatih_color//:color",
         "@com_github_spf13_cobra//:cobra",
         "@org_golang_x_term//:term",
+        "@org_golang_x_text//language",
+        "@org_golang_x_text//message",
     ],
 )
 

--- a/pkg/cmd/roachprod/main.go
+++ b/pkg/cmd/roachprod/main.go
@@ -31,8 +31,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/ui"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/vm"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
+	"github.com/fatih/color"
 	"github.com/spf13/cobra"
+	"golang.org/x/text/language"
+	"golang.org/x/text/message"
 )
 
 var rootCmd = &cobra.Command{
@@ -244,20 +248,26 @@ hosts file.
 		if listJSON && listDetails {
 			return errors.New("'json' option cannot be combined with 'details' option")
 		}
-		filteredCloud, err := roachprod.List(config.Logger, listMine, listPattern)
+		filteredCloud, err := roachprod.List(config.Logger, listMine, listPattern, vm.ListOptions{ComputeEstimatedCost: true})
+
 		if err != nil {
 			return err
 		}
 
 		// sort by cluster names for stable output.
 		names := make([]string, len(filteredCloud.Clusters))
+		maxClusterName := 0
 		i := 0
 		for name := range filteredCloud.Clusters {
 			names[i] = name
+			if len(name) > maxClusterName {
+				maxClusterName = len(name)
+			}
 			i++
 		}
 		sort.Strings(names)
 
+		p := message.NewPrinter(language.English)
 		if listJSON {
 			enc := json.NewEncoder(os.Stdout)
 			enc.SetIndent("", "  ")
@@ -266,15 +276,65 @@ hosts file.
 			}
 		} else {
 			// Align columns left and separate with at least two spaces.
-			tw := tabwriter.NewWriter(os.Stdout, 0, 8, 2, ' ', 0)
+			tw := tabwriter.NewWriter(os.Stdout, 0, 8, 2, ' ', tabwriter.AlignRight)
+			// N.B. colors use escape codes which don't play nice with tabwriter [1].
+			// We use a hacky workaround below to color the empty string.
+			// [1] https://github.com/golang/go/issues/12073
+
+			// Print header.
+			fmt.Fprintf(tw, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t\n",
+				"Cluster", "Clouds", "Size", "VM",
+				color.HiWhiteString("$/hour"), color.HiWhiteString("$ Spent"),
+				color.HiWhiteString("Uptime"), color.HiWhiteString("TTL"),
+				color.HiWhiteString("$/TTL"))
+			// Print separator.
+			fmt.Fprintf(tw, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t\n",
+				"", "", "",
+				color.HiWhiteString(""), color.HiWhiteString(""),
+				color.HiWhiteString(""), color.HiWhiteString(""),
+				color.HiWhiteString(""))
+			totalCostPerHour := 0.0
 			for _, name := range names {
 				c := filteredCloud.Clusters[name]
 				if listDetails {
 					c.PrintDetails(config.Logger)
 				} else {
-					fmt.Fprintf(tw, "%s\t%s\t%d", c.Name, c.Clouds(), len(c.VMs))
+					// N.B. Tabwriter doesn't support per-column alignment. It looks odd to have the cluster names right-aligned,
+					// so we make it left-aligned.
+					fmt.Fprintf(tw, "%s\t%s\t%d\t%s", name+strings.Repeat(" ", maxClusterName-len(name)), c.Clouds(),
+						len(c.VMs), c.VMs[0].MachineType)
 					if !c.IsLocal() {
-						fmt.Fprintf(tw, "\t(%s)", c.LifetimeRemaining().Round(time.Second))
+						colorByCostBucket := func(cost float64) func(string, ...interface{}) string {
+							switch {
+							case cost <= 100:
+								return color.HiGreenString
+							case cost <= 1000:
+								return color.HiBlueString
+							default:
+								return color.HiRedString
+							}
+						}
+						timeRemaining := c.LifetimeRemaining().Round(time.Second)
+						cost := c.CostPerHour
+						totalCostPerHour += cost
+						alive := timeutil.Since(c.CreatedAt).Round(time.Minute)
+						costSinceCreation := cost * float64(alive) / float64(time.Hour)
+						costRemaining := cost * float64(timeRemaining) / float64(time.Hour)
+						if cost > 0 {
+							fmt.Fprintf(tw, "\t%s\t%s\t%s\t%s\t%s\t",
+								color.HiGreenString(p.Sprintf("$%.2f", cost)),
+								colorByCostBucket(costSinceCreation)(p.Sprintf("$%.2f", costSinceCreation)),
+								color.HiWhiteString(alive.String()),
+								color.HiBlueString(timeRemaining.String()),
+								colorByCostBucket(costRemaining)(p.Sprintf("$%.2f", costRemaining)))
+						} else {
+							fmt.Fprintf(tw, "\t%s\t%s\t%s\t%s\t%s\t",
+								color.HiGreenString(""),
+								color.HiGreenString(""),
+								color.HiWhiteString(alive.String()),
+								color.HiBlueString(timeRemaining.String()),
+								color.HiGreenString(""))
+						}
 					} else {
 						fmt.Fprintf(tw, "\t(-)")
 					}
@@ -283,6 +343,10 @@ hosts file.
 			}
 			if err := tw.Flush(); err != nil {
 				return err
+			}
+
+			if totalCostPerHour > 0 {
+				_, _ = p.Printf("\nTotal cost per hour: $%.2f\n", totalCostPerHour)
 			}
 
 			// Optionally print any dangling instances with errors

--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -1088,7 +1088,7 @@ func (c *clusterImpl) validate(
 	// Perform validation on the existing cluster.
 	c.status("checking that existing cluster matches spec")
 	pattern := "^" + regexp.QuoteMeta(c.name) + "$"
-	cloudClusters, err := roachprod.List(l, false /* listMine */, pattern)
+	cloudClusters, err := roachprod.List(l, false /* listMine */, pattern, vm.ListOptions{})
 	if err != nil {
 		return err
 	}

--- a/pkg/roachprod/cloud/cluster_cloud.go
+++ b/pkg/roachprod/cloud/cluster_cloud.go
@@ -89,6 +89,9 @@ type Cluster struct {
 	CreatedAt time.Time     `json:"created_at"`
 	Lifetime  time.Duration `json:"lifetime"`
 	VMs       vm.List       `json:"vms"`
+	// CostPerHour is an estimate, in dollars, of how much this cluster costs to
+	// run per hour. 0 if the cost estimate is unavailable.
+	CostPerHour float64
 }
 
 // Clouds returns the names of all of the various cloud providers used
@@ -235,6 +238,7 @@ func ListCloud(l *logger.Logger, options vm.ListOptions) (*Cloud, error) {
 			if v.Lifetime < c.Lifetime {
 				c.Lifetime = v.Lifetime
 			}
+			c.CostPerHour += v.CostPerHour
 		}
 	}
 

--- a/pkg/roachprod/roachprod.go
+++ b/pkg/roachprod/roachprod.go
@@ -320,7 +320,9 @@ func Sync(l *logger.Logger, options vm.ListOptions) (*cloud.Cloud, error) {
 // List returns a cloud.Cloud struct of all roachprod clusters matching clusterNamePattern.
 // Alternatively, the 'listMine' option can be provided to get the clusters that are owned
 // by the current user.
-func List(l *logger.Logger, listMine bool, clusterNamePattern string) (cloud.Cloud, error) {
+func List(
+	l *logger.Logger, listMine bool, clusterNamePattern string, opts vm.ListOptions,
+) (cloud.Cloud, error) {
 	if err := LoadClusters(); err != nil {
 		return cloud.Cloud{}, err
 	}
@@ -344,7 +346,7 @@ func List(l *logger.Logger, listMine bool, clusterNamePattern string) (cloud.Clo
 		}
 	}
 
-	cld, err := Sync(l, vm.ListOptions{})
+	cld, err := Sync(l, opts)
 	if err != nil {
 		return cloud.Cloud{}, err
 	}

--- a/pkg/roachprod/vm/gce/BUILD.bazel
+++ b/pkg/roachprod/vm/gce/BUILD.bazel
@@ -17,6 +17,7 @@ go_library(
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_spf13_pflag//:pflag",
+        "@org_golang_google_api//cloudbilling/v1beta",
         "@org_golang_x_sync//errgroup",
     ],
 )

--- a/pkg/roachprod/vm/gce/gcloud.go
+++ b/pkg/roachprod/vm/gce/gcloud.go
@@ -31,6 +31,7 @@ import (
 	"github.com/cockroachdb/errors"
 	"github.com/spf13/pflag"
 	"golang.org/x/sync/errgroup"
+	cloudbilling "google.golang.org/api/cloudbilling/v1beta"
 )
 
 const (
@@ -175,22 +176,37 @@ func (jsonVM *jsonVM) toVM(
 	}
 
 	var volumes []vm.Volume
+	var localDisks []vm.Volume
+
+	parseDiskSize := func(size string) int {
+		if val, err := strconv.Atoi(size); err == nil {
+			return val
+		} else {
+			vmErrors = append(vmErrors, errors.Newf("invalid disk size: %q", size))
+		}
+		return 0
+	}
 
 	for _, jsonVMDisk := range jsonVM.Disks {
-		if jsonVMDisk.Source != "" && !jsonVMDisk.Boot {
+		if jsonVMDisk.Source == "" && jsonVMDisk.Type == "SCRATCH" {
+			// This is a scratch disk.
+			localDisks = append(localDisks, vm.Volume{
+				Size:               parseDiskSize(jsonVMDisk.DiskSizeGB),
+				ProviderVolumeType: "local-ssd",
+			})
+			continue
+		}
+		if !jsonVMDisk.Boot {
+			// Find a persistent volume (detailedDisk) matching the attached non-boot disk.
 			for _, detailedDisk := range disks {
 				if detailedDisk.SelfLink == jsonVMDisk.Source {
 					vol := vm.Volume{
 						ProviderResourceID: detailedDisk.Name,
-						ProviderVolumeType: jsonVMDisk.Type,
-						Zone:               jsonVM.Zone,
+						ProviderVolumeType: detailedDisk.Type,
+						Zone:               detailedDisk.Zone,
 						Name:               detailedDisk.Name,
 						Labels:             detailedDisk.Labels,
-					}
-					if val, err := strconv.Atoi(jsonVMDisk.DiskSizeGB); err == nil {
-						vol.Size = val
-					} else {
-						vmErrors = append(vmErrors, errors.Newf("invalid disk size: %q", jsonVMDisk.DiskSizeGB))
+						Size:               parseDiskSize(detailedDisk.SizeGB),
 					}
 					volumes = append(volumes, vol)
 				}
@@ -217,6 +233,7 @@ func (jsonVM *jsonVM) toVM(
 		SQLPort:                config.DefaultSQLPort,
 		AdminUIPort:            config.DefaultAdminUIPort,
 		NonBootAttachedVolumes: volumes,
+		LocalDisks:             localDisks,
 	}
 }
 
@@ -441,6 +458,9 @@ func (p *Provider) CreateVolume(
 }
 
 type instanceDisksResponse struct {
+	// Disks that are attached to the instance.
+	// N.B. Unattached disks can be enumerated via,
+	//	gcloud compute --project $project disks list --filter="-users:*"
 	Disks []attachDiskCmdDisk `json:"disks"`
 }
 type attachDiskCmdDisk struct {
@@ -1038,6 +1058,10 @@ func (p *Provider) FindActiveAccount(l *logger.Logger) (string, error) {
 
 // List queries gcloud to produce a list of VM info objects.
 func (p *Provider) List(l *logger.Logger, opts vm.ListOptions) (vm.List, error) {
+	if opts.IncludeVolumes {
+		l.Printf("WARN: --include-volumes is disabled; attached disks info will be partial")
+	}
+
 	var vms vm.List
 	for _, prj := range p.GetProjects() {
 		args := []string{"compute", "instances", "list", "--project", prj, "--format", "json"}
@@ -1079,11 +1103,156 @@ func (p *Provider) List(l *logger.Logger, opts vm.ListOptions) (vm.List, error) 
 		for _, jsonVM := range jsonVMS {
 			defaultOpts := p.CreateProviderOpts().(*ProviderOpts)
 			disks := userVMToDetailedDisk[jsonVM.SelfLink]
+
+			if len(disks) == 0 {
+				// Since `--include-volumes` is disabled, we convert attachDiskCmdDisk to describeVolumeCommandResponse.
+				// The former is a subset of the latter. Some information like `Labels` will be missing.
+				disks = toDescribeVolumeCommandResponse(jsonVM.Disks, jsonVM.Zone)
+			}
 			vms = append(vms, *jsonVM.toVM(prj, disks, defaultOpts))
 		}
 	}
 
+	if opts.ComputeEstimatedCost {
+		if err := populateCostPerHour(l, vms); err != nil {
+			// N.B. We continue despite the error since it doesn't prevent 'List' and other commands which may depend on it.
+
+			l.Errorf("Error during cost estimation (will continue without): %v", err)
+			if strings.Contains(err.Error(), "could not find default credentials") {
+				l.Printf("To fix this, run `gcloud auth application-default login`")
+			}
+		}
+	}
+
 	return vms, nil
+}
+
+// Convert attachDiskCmdDisk to describeVolumeCommandResponse and link via SelfLink, Source.
+func toDescribeVolumeCommandResponse(
+	disks []attachDiskCmdDisk, zone string,
+) []describeVolumeCommandResponse {
+	res := make([]describeVolumeCommandResponse, 0, len(disks))
+
+	diskType := func(s string) string {
+		if s == "PERSISTENT" {
+			// Unfortunately, we don't know if it's a pd-ssd or pd-standard. We assume pd_ssd since those are common.
+			return "pd-ssd"
+		}
+		return "unknown"
+	}
+
+	for _, d := range disks {
+		if d.Source == "" && d.Type == "SCRATCH" {
+			// Skip scratch disks.
+			continue
+		}
+		res = append(res, describeVolumeCommandResponse{
+			Name:     d.DeviceName,
+			SelfLink: d.Source,
+			SizeGB:   d.DiskSizeGB,
+			Type:     diskType(d.Type),
+			Zone:     zone,
+		})
+	}
+	return res
+}
+
+// populateCostPerHour adds an approximate cost per hour to each VM in the list,
+// using a basic estimation method.
+//  1. Compute and attached disks are estimated at the list prices, ignoring
+//     all discounts, but including any automatically applied credits.
+//  2. Boot disk costs are completely ignored.
+//  3. Network egress costs are completely ignored.
+//  4. Blob storage costs are completely ignored.
+func populateCostPerHour(l *logger.Logger, vms vm.List) error {
+	// Construct cost estimation service
+	ctx := context.Background()
+	service, err := cloudbilling.NewService(ctx)
+	if err != nil {
+		return err
+	}
+	beta := cloudbilling.NewV1betaService(service)
+	scenario := cloudbilling.EstimateCostScenarioWithListPriceRequest{
+		CostScenario: &cloudbilling.CostScenario{
+			ScenarioConfig: &cloudbilling.ScenarioConfig{
+				EstimateDuration: "3600s",
+			},
+			Workloads: []*cloudbilling.Workload{},
+		},
+	}
+	// Workload estimation service can handle 100 workloads at a time, so page
+	// 100 VMs at a time.
+	for len(vms) > 0 {
+		scenario.CostScenario.Workloads = scenario.CostScenario.Workloads[:0]
+		var page vm.List
+		if len(vms) <= 100 {
+			page, vms = vms, nil
+		} else {
+			page, vms = vms[:100], vms[100:]
+		}
+		for _, vm := range page {
+			machineType := vm.MachineType
+			zone := vm.Zone
+
+			workload := cloudbilling.Workload{
+				Name: vm.Name,
+				ComputeVmWorkload: &cloudbilling.ComputeVmWorkload{
+					InstancesRunning: &cloudbilling.Usage{
+						UsageRateTimeline: &cloudbilling.UsageRateTimeline{
+							UsageRateTimelineEntries: []*cloudbilling.UsageRateTimelineEntry{
+								{
+									// We're estimating the cost of 1 vm at a time.
+									UsageRate: 1,
+								},
+							},
+						},
+					},
+					MachineType: &cloudbilling.MachineType{
+						PredefinedMachineType: &cloudbilling.PredefinedMachineType{
+							MachineType: machineType,
+						},
+					},
+					PersistentDisks: []*cloudbilling.PersistentDisk{},
+					Region:          zone[:len(zone)-2],
+				},
+			}
+			for _, v := range vm.NonBootAttachedVolumes {
+				workload.ComputeVmWorkload.PersistentDisks = append(workload.ComputeVmWorkload.PersistentDisks, &cloudbilling.PersistentDisk{
+					DiskSize: &cloudbilling.Usage{
+						UsageRateTimeline: &cloudbilling.UsageRateTimeline{
+							Unit: "GiBy",
+							UsageRateTimelineEntries: []*cloudbilling.UsageRateTimelineEntry{
+								{
+									UsageRate: float64(v.Size),
+								},
+							},
+						},
+					},
+					DiskType: v.ProviderVolumeType,
+					Scope:    "SCOPE_ZONAL",
+				})
+			}
+			scenario.CostScenario.Workloads = append(scenario.CostScenario.Workloads, &workload)
+		}
+		estimate, err := beta.EstimateCostScenario(&scenario).Do()
+		if err != nil {
+			l.Errorf("Error estimating VM costs (will continue without): %v", err)
+			continue
+		}
+		workloadEstimates := estimate.CostEstimationResult.SegmentCostEstimates[0].WorkloadCostEstimates
+		for i := range workloadEstimates {
+			workloadEstimate := workloadEstimates[i].WorkloadTotalCostEstimate.NetCostEstimate
+			page[i].CostPerHour = float64(workloadEstimate.Units) + float64(workloadEstimate.Nanos)/1e9
+			// Add the estimated cost of local disks since the billing API only supports persistent disks.
+			for _, v := range page[i].LocalDisks {
+				// "For example, in the Iowa, Oregon, Taiwan, and Belgium regions, local SSDs cost $0.080 per GB per month."
+				// Normalized to per hour billing, 0.08 / 730 = 0.0001095890410958904
+				// https://cloud.google.com/compute/disks-image-pricing#disk
+				page[i].CostPerHour += float64(v.Size) * 0.00011
+			}
+		}
+	}
+	return nil
 }
 
 func serializeLabel(s string) string {

--- a/pkg/roachprod/vm/vm.go
+++ b/pkg/roachprod/vm/vm.go
@@ -90,8 +90,15 @@ type VM struct {
 	// LocalClusterName is only set for VMs in a local cluster.
 	LocalClusterName string `json:"local_cluster_name,omitempty"`
 
-	// NonBootAttachedVolumes are the non-bootable volumes attached to the VM.
+	// NonBootAttachedVolumes are the non-bootable, _persistent_ volumes attached to the VM.
 	NonBootAttachedVolumes []Volume `json:"non_bootable_volumes"`
+
+	// LocalDisks are the ephemeral SSD disks attached to the VM.
+	LocalDisks []Volume `json:"local_disks"`
+
+	// CostPerHour is the estimated cost per hour of this VM, in US dollars. 0 if
+	//there is no estimate available.
+	CostPerHour float64
 }
 
 // Name generates the name for the i'th node in a cluster.
@@ -209,6 +216,7 @@ func DefaultCreateOpts() CreateOpts {
 		GeoDistributed: false,
 		VMProviders:    []string{},
 		OsVolumeSize:   10,
+		CustomLabels:   map[string]string{"roachtest": "true"},
 	}
 	defaultCreateOpts.SSDOpts.UseLocalSSD = true
 	defaultCreateOpts.SSDOpts.NoExt4Barrier = true
@@ -270,7 +278,8 @@ type VolumeCreateOpts struct {
 }
 
 type ListOptions struct {
-	IncludeVolumes bool
+	IncludeVolumes       bool
+	ComputeEstimatedCost bool
 }
 
 // A Provider is a source of virtual machines running on some hosting platform.

